### PR TITLE
Fix Remote Nodes missing custom icons

### LINKED ISSUE #2084: Display custom class icons in the Remote scene tree view
### Describe the project you are working on

Card game.

### Describe the problem or limitation you are having in your project

Hard to read Remote view.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Custom class icons massively improve the readability of the scene tree, when it's used for logic (so mostly with scripts extending the Node class):

image

All that readability is lost in the Remote view though:

image

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The proposal is for the Remote view to display custom class icons in the same way the Local view does.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

It's core functionality.

### LINKED ISSUE #95209: Remote Scene doesn't use C# IconAttribute for node instanciated, but Inspector is able to use it?
### Tested versions

Probably reproductible on all version.

### System information

Window 10 - 4.3-RC2

### Issue description

I'm working with Godot C# and I use IconAttribute to have custom icon in the editor. `[GlobalClass, Icon("res://assets/textures/icons/tbd-icon.svg")]` Mostly everywhere you will be able to see the icon, but in the "Remote Scene" the tree view won't show any custom icon for instanciated node. You click on the node which update the Inspector view and you will see the EditorDebuggerRemoteObject do include the CustomIcon but not the Remote Scene 

image

### Steps to reproduce

1. This
```csharp
using System;
using Godot;

[GlobalClass, Icon("res://path/to/icon/some-icon.svg")]
public partial class GlobalManager : Node3D, IGlobalManager {
	public static GlobalManager Instance { get; private set; }
}
```
2. Make it Autoload or just instanciate in a working scene. 
3. Run the scene in the debugger and pause (or not!) and look at the Remote Scene tree where custom icon won't be use. 
4. For reference, I use this site (https://uxwing.com/) for icons, and edit the SVG for better representation of the icon. 

### Minimal reproduction project (MRP)

[mrp-custom-icon.zip](https://github.com/user-attachments/files/16514272/mrp-custom-icon.zip)

### PR DESCRIPTION
**Pull request by CrayolaEater**
# BEFORE:
image
# AFTER: 
image

This PR also fixes the type hint for Remote Nodes:
image

Closes https://github.com/godotengine/godot-proposals/issues/2084
Fixes #95209

### CODE REVIEW

File: scene/debugger/scene_debugger.cpp [hunk 1]
```diff
@@ -535,7 +535,22 @@ SceneDebuggerTree::SceneDebuggerTree(Node *p_root) {
...
+
+		StringName class_name = StringName();
+		ScriptInstance *script_instance = n->get_script_instance();
+		if (script_instance) {
+			Ref<Script> script = script_instance->get_script();
+			if (script.is_valid()) {
+				class_name = n->get_script_instance()->get_script()->get_global_name();
+
+				if (class_name == StringName())
+				{
```
Comment (AThousandShips) [inline scene/debugger/scene_debugger.cpp:547]:
```suggestion
				if (class_name == StringName()) {
```

File: editor/debugger/editor_debugger_tree.cpp [hunk 2]
```diff
@@ -32,6 +32,7 @@
 
 #include "editor/editor_node.h"
 #include "editor/gui/editor_file_dialog.h"
+#include "editor/editor_string_names.h"
```
Comment (Mickeon) [inline editor/debugger/editor_debugger_tree.cpp:35]: Static checks also complain about alphabetical order
```suggestion
#include "editor/editor_string_names.h"
#include "editor/gui/editor_file_dialog.h"
```

File: editor/debugger/editor_debugger_tree.cpp [hunk 3]
```diff
@@ -270,11 +275,14 @@ Variant EditorDebuggerTree::get_drag_data(const Point2 &p_point) {
...
 	String path = selected->get_text(0);
+	const int icon_size = get_theme_constant(SNAME("class_icon_size"), EditorStringName(Editor));
 
 	HBoxContainer *hb = memnew(HBoxContainer);
 	TextureRect *tf = memnew(TextureRect);
 	tf->set_texture(selected->get_icon(0));
-	tf->set_stretch_mode(TextureRect::STRETCH_KEEP_CENTERED);
+	tf->set_custom_minimum_size(Size2(icon_size, icon_size));
+	tf->set_stretch_mode(TextureRect::STRETCH_KEEP_ASPECT_CENTERED);
+	tf->set_expand_mode(TextureRect::EXPAND_IGNORE_SIZE);
```
Comment (MewPurPur) [inline editor/debugger/editor_debugger_tree.cpp:285]: In 4.2.1, this property is still marked as experimental, and I don't think this has changed in 4.3

Maybe it's time to reconsider?

Comment (AThousandShips) [reply editor/debugger/editor_debugger_tree.cpp:285]: According to the documentation it's specifically the fit modes that are experimental:
> Using EXPAND_FIT_WIDTH, EXPAND_FIT_WIDTH_PROPORTIONAL, EXPAND_FIT_HEIGHT or EXPAND_FIT_HEIGHT_PROPORTIONAL may result in unstable behavior in some containers

Though the property itself is marked experimental it is used in several places in the editor

File: scene/debugger/scene_debugger.cpp [hunk 4]
```diff
@@ -535,7 +535,21 @@ SceneDebuggerTree::SceneDebuggerTree(Node *p_root) {
 				}
 			}
 		}
-		nodes.push_back(RemoteNode(count, n->get_name(), n->get_class(), n->get_instance_id(), n->get_scene_file_path(), view_flags));
+
+		StringName class_name;
```
Comment (AThousandShips) [inline scene/debugger/scene_debugger.cpp:539]:
```suggestion
		String class_name;
```

File: scene/debugger/scene_debugger.cpp [hunk 5]
```diff
@@ -535,7 +535,21 @@ SceneDebuggerTree::SceneDebuggerTree(Node *p_root) {
...
+		ScriptInstance *script_instance = n->get_script_instance();
+		if (script_instance) {
+			Ref<Script> script = script_instance->get_script();
+			if (script.is_valid()) {
+				class_name = script->get_global_name();
+
+				if (class_name == StringName()) {
```
Comment (AThousandShips) [inline scene/debugger/scene_debugger.cpp:546]:
```suggestion
				if (class_name.is_empty()) {
```

File: scene/debugger/scene_debugger.cpp [hunk 6]
```diff
@@ -535,7 +535,21 @@ SceneDebuggerTree::SceneDebuggerTree(Node *p_root) {
...
+					// If there is no class_name in this script we just take the script path.
+					class_name = script->get_path();
+				}
+			}
+		}
+		nodes.push_back(RemoteNode(count, n->get_name(), class_name != StringName() ? class_name : n->get_class(), n->get_instance_id(), n->get_scene_file_path(), view_flags));
```
Comment (AThousandShips) [inline scene/debugger/scene_debugger.cpp:552]:
```suggestion
		nodes.push_back(RemoteNode(count, n->get_name(), class_name.is_empty() ? n->get_class() : class_name, n->get_instance_id(), n->get_scene_file_path(), view_flags));
```

**Review by AThousandShips — COMMENTED** [review]
No reason to use `StringName` here when one of the cases is a `String` and the method takes a `String`

**Comment by Mickeon** [comment]
I approve of this but I cannot speak on behalf of the debugger team.

It's unfortunate that the Remote Tree is a completely different thing from the opened Scene display, because code duplication like the one in the PR is required. I considered merging the two someday but it's an very, very daunting task.

**Comment by jaydensipe** [comment]
This will close https://github.com/godotengine/godot-proposals/issues/2084 and #95209 when merged.

File: editor/debugger/editor_debugger_tree.cpp [hunk 7]
```diff
@@ -63,7 +63,6 @@ void EditorDebuggerTree::_notification(int p_what) {
 			connect("item_collapsed", callable_mp(this, &EditorDebuggerTree::_scene_tree_folded));
 			connect("item_mouse_selected", callable_mp(this, &EditorDebuggerTree::_scene_tree_rmb_selected));
 		} break;
-
```
Comment (KoBeWi) [inline editor/debugger/editor_debugger_tree.cpp:66]: Why this change?

**Review by KoBeWi — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

**Pull request merged**